runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/OutputDeduplicatorTest.java

```diff
@@ -0,0 +1,438 @@
+    RunnerApi.Components components =
+        Components.newBuilder()
+            .putTransforms("one", one)
+            .putPcollections("one.out", oneOut)
+            .putTransforms("two", two)
+            .putPcollections("two.out", twoOut)
+            .putTransforms("red", red)
+            .putPcollections("red.out", redOut)
+            .putTransforms("blue", blue)
+            .putPcollections("blue.out", blueOut)
+            .build();
+    ExecutableStage oneStage =
+        ImmutableExecutableStage.of(
+            components,
+            Environment.getDefaultInstance(),
+            PipelineNode.pCollection("red.out", redOut),
+            ImmutableList.of(),
+            ImmutableList.of(PipelineNode.pTransform("one", one)),
+            ImmutableList.of(PipelineNode.pCollection("one.out", oneOut)));
+    ExecutableStage twoStage =
+        ImmutableExecutableStage.of(
+            components,
+            Environment.getDefaultInstance(),
+            PipelineNode.pCollection("red.out", redOut),
+            ImmutableList.of(),
+            ImmutableList.of(PipelineNode.pTransform("two", two)),
+            ImmutableList.of(PipelineNode.pCollection("two.out", twoOut)));
+    PTransformNode redTransform = PipelineNode.pTransform("red", red);
+    PTransformNode blueTransform = PipelineNode.pTransform("blue", blue);
+    QueryablePipeline pipeline = QueryablePipeline.forPrimitivesIn(components);
+    DeduplicationResult result =
+        OutputDeduplicator.ensureSingleProducer(
+            pipeline,
+            ImmutableList.of(oneStage, twoStage),
+            ImmutableList.of(redTransform, blueTransform));
+
+    assertThat(result.getDeduplicatedComponents(), equalTo(components));
+    assertThat(result.getDeduplicatedStages().keySet(), empty());
+    assertThat(result.getDeduplicatedTransforms().keySet(), empty());
+    assertThat(result.getIntroducedTransforms(), empty());
+  }
+
+  @Test
+  public void duplicateOverStages() {
+    /* When multiple stages and a runner-executed transform produce a PCollection, all should be
+     * replaced with synthetic flattens.
+     * S -> A; T -> A becomes S -> A'; T -> A''; A', A'' -> Flatten -> A
+     */
+    PTransform one =
+        PTransform.newBuilder().putInputs("in", "red.out").putOutputs("out", "one.out").build();
+    PCollection oneOut = PCollection.newBuilder().setUniqueName("one.out").build();
+    PTransform two =
+        PTransform.newBuilder().putInputs("in", "red.out").putOutputs("out", "two.out").build();
+    PCollection twoOut = PCollection.newBuilder().setUniqueName("two.out").build();
+    PCollection sharedOut = PCollection.newBuilder().setUniqueName("shared.out").build();
+    PTransform shared =
+        PTransform.newBuilder()
+            .putInputs("one", "one.out")
+            .putInputs("two", "two.out")
+            .putOutputs("shared", sharedOut.getUniqueName())
```

Comment:
Nit: for consistency, why are you using a literal in some places, and getUniqueName in others?